### PR TITLE
Rework the native plugin builder

### PR DESCRIPTION
**Pull request by swift-kim**
Implements #193.

- Each plugin is built as a static library, and included when the final shared object (`libflutter_plugins.so`) is built.
- Because the target project type is now `staticLib`, linker-specific properties defined in `project_def.prop` do not apply. However, you can still place your libs in `lib` or arch-specific directories under it, such as `lib/arm` and `lib/armel`. The libs will be automatically included during linking even though they are not defined anywhere.
  - `lib/arm/libstatic.a` precedes `lib/libstatic.a` if they both exist.

cc @WonyoungChoi

The inputs and outputs of the `tizen_native_plugins` target include:

(Inputs)
- `flutter-tizen/lib/build_targets/plugins.dart`
- `flutter-tizen/lib/tizen_sdk.dart`
- `APP_DIR/.packages` (to create a list of plugins to build)
- `APP_DIR/tizen/tizen-manifest.xml` (to check the target profile and API version)
- `flutter/bin/cache/artifacts/engine/tizen-common/cpp_client_wrapper/**` 
- `flutter/bin/cache/artifacts/engine/tizen-common/public/**`
- `flutter/bin/cache/artifacts/engine/tizen-ARCH-MODE/libflutter_tizen_PROFILE.so`
- `PLUGIN_DIR/tizen/project_def.prop`
- `PLUGIN_DIR/tizen/inc/**`
- `PLUGIN_DIR/tizen/src/**`
- `PLUGIN_DIR/tizen/lib/**`

(Outputs)
- `BUILD_DIR/tizen_plugins/lib/*.so` (copied from `PLUGIN_DIR/tizen/lib/`)
- `BUILD_DIR/tizen_plugins/include/*.h` (copied from `PLUGIN_DIR/tizen/inc/PLUGIN_CLASS.h`)
- `BUILD_DIR/tizen_plugins/libflutter_plugins.so` (build output)

### CODE REVIEW

File: templates/plugin/cpp/project_def.prop.tmpl [hunk 1]
```diff
@@ -2,7 +2,7 @@
 # for details.
 
 APPNAME = {{projectName}}_plugin
-type = sharedLib
+type = staticLib
```
Comment (swift-kim) [inline templates/plugin/cpp/project_def.prop.tmpl:5]: Existing plugins are not affected. The builder always uses `staticLib` as a project type when building plugins.

**Comment by WonyoungChoi** [comment]
@swift-kim It feels like the build function is too long. How about separating the part that builds the static library into separate targets or private methods?

**Comment by swift-kim** [comment]
@WonyoungChoi Splitting into multiple targets will make it hard to manage input and output dependencies.

Separating methods looks like a better option, but it will make the code much longer as many variables are shared between the two parts. They are closed related to each other.

**Comment by WonyoungChoi** [comment]
@swift-kim Well.. I have a slightly different view. I think separating the target will make the input/output dependencies more clear. 
To put it in the pseudo Makefile grammar, 
```makefile
sharedlibrary: staticlibraryA, staticLibraryB
    // generating the shared library with the generated static libraries
staticLibraryA: headerfiles of packageA, sourcefiles of packageA
    // generating the static library with files of packageA
staticLibraryB: headerfiles of packageB, sourcefiles of packageB
    // generating the static library with files of packageB
```
The current implementation is that the main target uses all dependent source files as input.

**Comment by swift-kim** [comment]
@WonyoungChoi That's a too broad view. In reality, there are more inputs and outputs than that, especially shared libs and static libs provided by each plugin.

File: lib/tizen_plugins.dart [hunk 2]
```diff
@@ -66,6 +66,16 @@ class TizenPlugin extends PluginPlatform implements NativeOrDartPlugin {
   final String dartPluginClass;
   final String fileName;
 
+  TizenPlugin copyWith({Directory directory}) {
```
Comment (WonyoungChoi) [inline lib/tizen_plugins.dart:69]: How about changing to `copyTo` and including the directory copying. 
```dart
TizenPlugin copyTo(Directory destination) {
  if (destination == null) {
    throw Exception('destination is null');
  }
  copyDirectory(directory, destination);
  return TizenPlugin(
    ....
}
```

File: lib/build_targets/plugins.dart [hunk 3]
```diff
@@ -64,151 +64,206 @@ class NativePlugins extends Target {
...
+      );
+      copyDirectory(plugin.directory, pluginCopy.directory);
+
+      final List<String> properties = <String>[];
+      for (String line in plugin.projectFile.readAsLinesSync()) {
+        if (line.startsWith('type =')) {
+          line = 'type = staticLib';
+        }
+        properties.add(line);
+      }
```
Comment (WonyoungChoi) [inline lib/build_targets/plugins.dart:130]: maybe this part can be extracted to `_ensureStaticLibType(plugin)`. 
And I the type is not `staticLib`, I think the information that is has been changed should be notified as some log message.

File: lib/tizen_plugins.dart (hunk 2 above)
Comment (swift-kim) [reply lib/tizen_plugins.dart:69]: Actually it was an adaptation of the "copyWith" pattern widely used when developing Flutter apps.

File: lib/build_targets/plugins.dart (hunk 3 above)
Comment (swift-kim) [reply lib/build_targets/plugins.dart:130]: Nice idea!

**Comment by swift-kim** [comment]
I agree that the build method is complicated and hard to follow the overall flow at a glance. It's purposefully complicated, because a lot of input and output files involve during the build. I'll provide some map of inputs and outputs associated with the target if you need.

File: lib/build_targets/plugins.dart [hunk 4]
```diff
@@ -64,151 +64,206 @@ class NativePlugins extends Target {
...
+        '-fPIC',
+        '-I"${clientWrapperDir.childDirectory('include').path.toPosixPath()}"',
+        '-I"${publicDir.path.toPosixPath()}"',
+        '-D${buildInfo.deviceProfile.toUpperCase()}_PROFILE',
+      ];
 
-      final Directory headerDir = plugin.directory.childDirectory('inc');
-      if (headerDir.existsSync()) {
-        headerDir
+      final RunResult result = await _processUtils.run(<String>[
```
Comment (WonyoungChoi) [inline lib/build_targets/plugins.dart:144]: How about adding `buildNative()` method to `TizenSdk`?
```dart
Future<File> buildNative(
    TizenBuildInfo buildInfo,
    Rootstrap rootstrap, 
    Directory buildDirectory,
    String expectedOutputPath, { 
    List<String> extras,
    Map<String, String> environment,
    }) async {
      ....
```
this part can be more simple like below:
```dart
     final File libFile = await tizenSdk.buildNative(
        buildInfo,
        rootstrap,
        pluginCopy.directory,
        'lib$libName.a',
        extras: <String>[
          '-fPIC',
          '-I"${clientWrapperDir.childDirectory('include').path.toPosixPath()}"',
          '-I"${publicDir.path.toPosixPath()}"',
          '-D${buildInfo.deviceProfile.toUpperCase()}_PROFILE',
        ],
        environment: <String, String>{
          'PATH': getDefaultPathVariable(),
        },
      );
```

Comment (swift-kim) [reply lib/build_targets/plugins.dart]: Thanks! I applied the suggestion but modified the method signature a bit to match the actual parameters of the `tizen build-native` command.

Comment (swift-kim) [reply lib/build_targets/plugins.dart]: I forgot to apply this to the NativeTpk target in `package.dart`. I'll update it as well.

File: lib/tizen_sdk.dart [hunk 5]
```diff
@@ -116,6 +122,36 @@ class TizenSdk {
...
+    String predefine,
+    List<String> extraOptions = const <String>[],
+    Rootstrap rootstrap,
+    Map<String, String> environment,
+  }) async {
+    assert(configuration != null);
+    assert(arch != null);
+
+    return _processUtils.run(<String>[
+      tizenSdk.tizenCli.path,
```
Comment (WonyoungChoi) [inline lib/tizen_sdk.dart:139]:
```suggestion
      tizenCli.path,
```

File: lib/tizen_sdk.dart [hunk 6]
```diff
@@ -116,6 +122,36 @@ class TizenSdk {
...
+      'build-native',
+      '-C',
+      configuration,
+      '-a',
+      arch,
+      '-c',
+      compiler ?? defaultNativeCompiler,
+      if (predefine != null) ...<String>['-d', predefine],
```
Comment (WonyoungChoi) [inline lib/tizen_sdk.dart:147]: any cases where multiple predefine values come in?

Comment (swift-kim) [reply lib/tizen_sdk.dart:147]: I tested
```
tizen build-native -d "FIRST_DEFINE SECOND_DEFINE THIRD_DEFINE" ...
```
but it didn't work. The predefine option expects only one argument. If you have more than one user defines, they should go in `extraOptions`.

Comment (swift-kim) [reply lib/tizen_sdk.dart:147]:
```
tizen build-native -d FIRST_DEFINE -d SECOND_DEFINE -d THIRD_DEFINE ...
```
This worked. I'll make a change using `for`.

**Review by WonyoungChoi — APPROVED** [review]
LGTM 
👍 Thank you for your hard work. :)

**Review by bbrto21 — APPROVED** [review]
